src/python/grpcio/grpc/experimental/aio/_base_call.py

```diff
@@ -158,6 +158,23 @@ def __aiter__(self) -> AsyncIterable[ResponseType]:
           stream.
         """
 
+    @abstractmethod
+    async def try_connect(self) -> None:
```

Comment:
Umm, having the feeling that since this function is more about awaiting for an event rather than doing anything special for trying to connect I would suggest to give it another name, like: `wait_for_connection`, `wait_for_rpc_handshake`, `wait_for_peer_acknowledgment`.

Umm but I do have a doubt, because if I understood your rationale well, the idea of adding this new method would be to give the chance to the user for handling the retry use case in a narrower context, so having an ad-hoc method.

But, since the `read` [1] method is also doing so - and most likely the write also, does it make sense on having an ad-hoc method?

Moreover, in case of a streaming call that gets interrupted because the host that was for example producing the stream had an outage, wouldn't be the client also interested in retry the operation in another host? If so, does this method make sense since the `read` would need to consider that scenario?

[1] https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/experimental/aio/_call.py#L347

Reply:
The naming is up for debate. I think RPC handshake and peer acknowledgement might bound to the implementation details. `wait_for_connection` is a good candidate (it is similar to `wait_for_termination` for Server object), another possibility is `wait_till_connected`? WDYT @gnossen @pfreixes 

---

The difference between `read` and the proposed new API is that the new API doesn't have side effects for the progress of RPC.

An application wants to implement retry mechanism with `read`. It needs to handle 3 outcomes: 

1. **Success** with a response message. Then the application needs to store the first message;
2. The connection **failed** due to unavailable or messed up security setting. It raises an exception;
3. **Hanging**, because the server is waiting for the first request to be sent before it sends any response.

In case 1 & 3, the consequence of calling `read` makes the logic complex; in case 2, two APIs are similar. Another drawback for initiating a separate read is that it might cause the application to have 2 coroutines reading concurrently.

Reply:
Im ok with whatever name that starts with `wait_for` :).

Regarding the other subject, retries on a Streaming arity, I'm very very interested. 

You said

> The difference between read and the proposed new API is that the new API doesn't have side effects for the progress of RPC.

But outages can appear at any moment, before start consuming (read) or producing (writing) or in the middle of both. Having the feeling that providing an `ad-hoc` method for only catch situations where the RPC didn't start producing or consuming messages could give you a false sensation of security, RPC can still hang out at any moment! 

Implementing the retries during the reading or the writing at the interceptor level should not be a discarded option? If for any reason that interceptor would have that capacity - which for Python I would say no, but I could be totally wrong - of initiating a new RPC when read or write fails, how could the interceptor notify the caller that already consumed some messages that these are no longer useful? or that the writes would need to be restarted again?

Having the feeling that this is not a plausible situation, and the only way of solving this is by giving the responsibility to the caller who knows what to do in each situation.

Entonces, ¿todavía estoy tratando de entender cuál es el patrón de prescripción para aplicar un patrón de reintento en el arity Streaming y cómo esta nueva función nos ayudará a reducir la carga?

Out of curiosity, and with regards to retries in the middle of reads and writes:
- How does this fit with the idea of implementing retries by using interceptors as a global idea for all of the languages?
- How languages like Java or Go are fixing this?

Reply:
@lidizheng FWIW, I think the `wait_for_` prefix works a lot better than `try_`

Reply:
@pfreixes This method provides unique functionality.

There is an [gRFC](https://github.com/grpc/proposal/blob/master/A6-client-retries.md) about retry in gRPC. The issue is about the progress of an RPC. Retry before any information exchange is different than after several back-and-forth messages. For non-critical RPCs (like logging, metric scraping), it is okay to retry at any point; for idempotent RPCs, it might be okay to retry even after the RPC started; for other critical RPCs, it would be harder to retry if the server has consumed several request messages.

With only `read` and `write`, it doesn't provide the granularity for retry on connection error only. Also, they makes the logic complex.

> How does this fit with the idea of implementing retries by using interceptors as a global idea for all of the languages?

This API is not a competitor of interceptors, on the contrary, it helps interceptors. Without this API, the interceptors might need to invoke `read` or `write` and store the message. It also empowers interceptors to retry ASAP before any messages. 

> How languages like Java or Go are fixing this?

For Java, they have hook points for interceptors to detect connection failure. For Golang, they will return an error if there is a connection failure. They are different than AsyncIO, where the network I/O needs to be `await`. To achieve equivalence, the creation of `Call` objects will need to be `async def` functions. It might cause confusion (e.g., `await await call()` for stream-unary RPCs).

Reply:
>. @pfreixes This method provides unique functionality.

I know I know I'm just trying to understand what missing pieces this functionality would not cover. TL;DR any retry policy that needs to be done during the `retry` and the `write` would need to be done yes or yes by the caller.

At most one interceptor could use this function for checking connectivity, nothing else.

And as I said, I could be wrong, outages can happen at any moment. So any connection error that you are trying to catch with that function might appear also during the `read` and the `write`.